File: apps/api/app/schemas/audit.py
import uuid
from datetime import datetime

from pydantic import BaseModel, HttpUrl, field_validator


class AuditCreate(BaseModel):
    url: HttpUrl
    project_id: uuid.UUID | None = None
    max_pages: int = 500

    @field_validator("max_pages")
    @classmethod
    def validate_max_pages(cls, v: int) -> int:
        if v < 1 or v > 10000:
            raise ValueError("max_pages must be between 1 and 10,000")
        return v


class AuditResponse(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID | None
    url: str
    status: str
    status_message: str | None
    pages_crawled: int
    max_pages: int
    score_performance: float | None
    score_accessibility: float | None
    score_best_practices: float | None
    score_seo: float | None
    lcp_ms: float | None
    inp_ms: float | None
    cls: float | None
    celery_task_id: str | None
    created_at: datetime
    completed_at: datetime | None

    model_config = {"from_attributes": True}


class AuditListResponse(BaseModel):
    id: uuid.UUID
    url: str
    status: str
    pages_crawled: int
    score_seo: float | None
    created_at: datetime

    model_config = {"from_attributes": True}

File: apps/web/src/types/audit.ts
export type AuditStatus = "pending" | "crawling" | "analyzing" | "completed" | "failed";
export type Severity = "critical" | "error" | "warning" | "info";
export type CWVRating = "good" | "needs-improvement" | "poor";

export interface Audit {
  id: string;
  url: string;
  status: AuditStatus;
  status_message: string | null;
  project_id: string;
  pages_crawled: number;
  max_pages: number;
  score_performance: number | null;
  score_accessibility: number | null;
  score_best_practices: number | null;
  score_seo: number | null;
  lcp_ms: number | null;
  inp_ms: number | null;
  cls: number | null;
  desktop_scores?: {
    performance: number | null;
    accessibility: number | null;
    best_practices: number | null;
    seo: number | null;
    lcp_ms: number | null;
    inp_ms: number | null;
    cls: number | null;
  } | null;
  celery_task_id: string | null;
  created_at: string;
  completed_at: string | null;
}

export interface SEOIssue {
  id: string;
  page_url: string;
  severity: Severity;
  category: string;
  message: string;
  pages_affected?: number;
}

export interface CoreWebVitals {
  lcp_ms: number | null;
  inp_ms: number | null;
  cls: number | null;
  lcp_rating: CWVRating | null;
  inp_rating: CWVRating | null;
  cls_rating: CWVRating | null;
}

export interface SiteStructure {
  total_pages: number;
  max_depth: number;
  broken_links: number;
  redirect_chains: number;
  orphan_pages: number;
  avg_load_time_ms: number;
}

export interface AuditDetail extends Audit {
  issues: SEOIssue[];
  core_web_vitals: CoreWebVitals;
  site_structure: SiteStructure;
}
